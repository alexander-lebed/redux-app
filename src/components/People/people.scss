@import '../../../css/variables';

.search-container {
  padding-bottom: 5px;
  align-items: baseline;
  color: $secondary-color;
}

.search-input {
  font-size: 16px !important;
  border: none !important;
  box-shadow: none !important;
}

.last-seen {
  color: $secondary-color;
  font-size: 13px;
}