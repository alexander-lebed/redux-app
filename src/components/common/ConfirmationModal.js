// @flow
import React from 'react'
import { useSelector } from 'react-redux';
import Modal from 'react-bootstrap/Modal'
import Button from 'react-bootstrap/Button';
import type { Node } from 'react';
import type { Translation } from '../../types';

type Props = {
    title: Node | string,
    body: Node | string,
    onConfirm: Function,
    onCancel: Function,
}

const ConfirmationModal = (props: Props) => {
    const translation: Translation = useSelector(state => state.translation);
    return (
        <Modal show onHide={props.onCancel}>
            <Modal.Header >
                <Modal.Title>
                    {props.title}
                </Modal.Title>
            </Modal.Header>
            <Modal.Body>
                {props.body}
            </Modal.Body>
            <Modal.Footer>
                <Button
                    variant='success'
                    size='sm'
                    onClick={props.onConfirm}
                >
                    {translation.COMMON.YES}
                </Button>
                <Button
                    variant='outline-dark'
                    size='sm'
                    onClick={props.onCancel}
                >
                    {translation.COMMON.CANCEL}
                </Button>
            </Modal.Footer>
        </Modal>
    )
};

export default ConfirmationModal;
