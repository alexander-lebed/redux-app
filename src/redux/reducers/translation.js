// @flow
import moment from 'moment';
import { setLocale } from './locale';
import enStrings from '../../lang/en';
import ruStrings from '../../lang/ru';
import type { Action, Dispatch, Locale, Translation } from '../../types';

const actions = {
    SET_STRINGS: 'SET_STRINGS'
};

const strings = (state: Translation = enStrings, action: Action) => {
    switch (action.type) {
        case actions.SET_STRINGS: {
            return {...action.payload};
        }
        default:
            return state;
    }
};

export default strings;

export function initTranslation() {
    return (dispatch: Dispatch, getState: Function) => {
        dispatch(translate(getState().locale));
    }
}

export function translate(locale: Locale) {
    return (dispatch: Dispatch) => {

        dispatch(setLocale(locale));

        moment.locale(locale);

        if (locale === 'ru') {
            dispatch({
                type: actions.SET_STRINGS,
                payload: ruStrings
            })
        } else {
            dispatch({
                type: actions.SET_STRINGS,
                payload: enStrings
            })
        }
    }
}