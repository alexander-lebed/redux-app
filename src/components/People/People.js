// @flow
import React from 'react';
import { connect, useSelector } from 'react-redux';
import queryString from 'query-string';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import Col from 'react-bootstrap/Col';
import Table from 'react-bootstrap/Table';
import Form from 'react-bootstrap/Form';
import Modal from 'react-bootstrap/Modal';
import InputGroup from 'react-bootstrap/InputGroup';
import Button from 'react-bootstrap/Button';
import Image from 'react-bootstrap/Image';
import Jumbotron from 'react-bootstrap/Jumbotron';
import { ONLINE_STYLE, adminEmail } from '../../constants';
import { timestampToHumanDate } from '../../utils';
import { deleteUser } from '../../redux/reducers/users';
import ConfirmationModal from '../common/ConfirmationModal';
import type { User, Translation } from '../../types';

type Props = {
    history: Object,
    user: User,
    users: Array<User>,
    translation: Translation,
    deleteUser: Function
}

type State = {
    searchText: string,
    deleteUserId: string,
    clickedUser: User | null
}

export class People extends React.Component<Props, State> {

    state: State;

    constructor(props: Props) {
        super(props);
        this.state = {
            searchText: '',
            deleteUserId: '',
            clickedUser: null
        }
    }

    showUserProfile = (user: User | null) => this.setState({clickedUser: user});

    showDeleteConfirmation = (userId: string) => this.setState({deleteUserId: userId});

    hideDeleteConfirmation = () => this.setState({deleteUserId: ''});

    deleteUser = () => {
        this.props.deleteUser(this.state.deleteUserId);
        this.hideDeleteConfirmation();
    };

    render() {
        let {users, history, translation} = this.props;
        const {PEOPLE, COMMON} = translation;
        if (this.state.searchText) {
            users = users.filter(e => e.username.toLowerCase().indexOf(this.state.searchText.toLowerCase()) !== -1)
        }
        return (
            <Container fluid>
                <Row>
                    <Col md={{span: 10, offset: 1}} lg={{span: 8, offset: 2}}>
                        <Row>
                            <Col xs={12} sm={6}>
                                <InputGroup size='sm' className='search-container'>
                                    <InputGroup.Prepend>
                                        <i className='fas fa-search' />
                                    </InputGroup.Prepend>
                                    <Form.Control
                                        placeholder={PEOPLE.SEARCH_PEOPLE}
                                        aria-label='Search people'
                                        aria-describedby='search'
                                        className='search-input'
                                        value={this.state.searchText}
                                        onChange={e => this.setState({searchText: e.target.value})}
                                    />
                                </InputGroup>
                            </Col>
                        </Row>
                        {users.length === 0
                            ?  <Jumbotron className='text-center'>{COMMON.NO_RESULTS}</Jumbotron>
                            : <Table>
                                <tbody>
                                    {users.map(user => (
                                        <UserRow
                                            key={user._id}
                                            history={history}
                                            user={user}
                                            showUserProfile={u => this.showUserProfile(u)}
                                            showDeleteConfirmation={userId => this.showDeleteConfirmation(userId)}
                                        />
                                    ))}
                                </tbody>
                            </Table>
                        }
                        {this.state.clickedUser &&
                        <Modal
                            show={this.state.clickedUser !== null}
                            className='profile-modal'
                            onHide={() => this.showUserProfile(null)}
                        >
                            <Modal.Header closeButton>
                                <Modal.Title>{this.state.clickedUser.username}</Modal.Title>
                            </Modal.Header>
                            <Modal.Body style={{textAlign: 'center'}}>
                                <Image
                                    className='profile-modal-picture'
                                    src={this.state.clickedUser.pictureUrl ? this.state.clickedUser.pictureUrl : '/images/default-profile.jpg'}
                                />
                            </Modal.Body>
                        </Modal>
                        }
                        {this.state.deleteUserId &&
                        <ConfirmationModal
                            title={COMMON.DELETE_CONFIRMATION}
                            body={PEOPLE.DELETE_CONFIRMATION}
                            onConfirm={() => this.deleteUser()}
                            onCancel={() => this.hideDeleteConfirmation()}
                        />
                        }
                    </Col>
                </Row>
            </Container>
        )
    }
}

export default connect(
    state => ({
        user: state.authentication.user,
        users: state.users.users,
        translation: state.translation
    }),
    { deleteUser }
)(People);


type UserRowProps = {
    history: Object,
    user: User,
    showUserProfile: Function,
    showDeleteConfirmation: Function
}
export const UserRow = ((props: UserRowProps) => {
    const {user, showUserProfile, showDeleteConfirmation, history} = props;
    const {loggedUser, translation} = useSelector(state => ({
        loggedUser: state.authentication.user,
        translation: state.translation
    }));

    const isAdmin = loggedUser.email === adminEmail;

    const imageStyle = user.online ? ONLINE_STYLE : {};

    const goToConversationWith = (userIds: Array<string>) => {
        const query = queryString.stringify({userIds});
        history.push(`/conversation?${query}`);
    };
    return (
        <tr key={user._id} id={user._id}>
            <td style={{padding: 10}}>
                <Row>
                    <Col xs={12} sm={6}>
                        <div
                            className='profile-picture-wrapper cursor'
                            onClick={() => showUserProfile(user)}
                        >
                            <Image
                                id={`user-picture-${user._id}`}
                                roundedCircle
                                style={imageStyle}
                                className='profile-picture'
                                src={user.pictureUrl ? user.pictureUrl : '/images/default-profile.jpg'}
                            />
                        </div>
                        <div>
                            {user.username}
                            <div className='last-seen'>
                                {!user.online && `${translation.PEOPLE.LAST_SEEN} ${timestampToHumanDate(user.lastTime, false, translation)}`}
                            </div>
                        </div>
                    </Col>
                    <Col xs={12} sm={6}>
                        <div className='pull-right pt-1 pt-sm-2'>
                            {isAdmin &&
                            <>
                                <Button
                                    id={`delete-user-${user._id}`}
                                    size='sm'
                                    variant='outline-danger'
                                    onClick={() => showDeleteConfirmation(user._id)}
                                >
                                    <i className='far fa-trash-alt' style={{marginRight: 5}} /> {translation.COMMON.DELETE}
                                </Button>
                                {' '}
                            </>
                            }
                            <Button
                                id={`write-user-${user._id}`}
                                size='sm'
                                variant='outline-dark'
                                onClick={() => goToConversationWith([user._id])}
                            >
                                <i className='far fa-paper-plane' style={{marginRight: 5}} /> {translation.PEOPLE.WRITE_MESSAGE}
                            </Button>
                        </div>
                    </Col>
                </Row>
            </td>
        </tr>
    )
});
UserRow.displayName = 'UserRow';