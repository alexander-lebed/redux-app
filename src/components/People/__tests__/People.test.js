// @flow
import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import { People, UserRow } from '../People';
import ConfirmationModal from '../../common/ConfirmationModal';
import translation from '../../../lang/en';

jest.mock('react-redux', () => ({
    connect: (mapStateToProps, mapDispatchToProps) => (ReactComponent) => ({
        mapStateToProps,
        mapDispatchToProps,
        ReactComponent
    }),
    useSelector: () => ({
        loggedUser: {email: 'alexanderlebed999@gmail.com'},
        translation: {PEOPLE: {}, COMMON: {}},
        COMMON: {}
    })
}));

describe('<People/>', () => {
    const props = {
        user: {
            _id: '111',
            username: 'Current User',
            email: 'alexanderlebed999@gmail.com',
            password: 'password',
            pictureUrl: '',
            online: true,
            lastTime: 1519294933743,
            oauth: ''
        },
        users: [
            {
                _id: '111',
                username: 'Current User',
                email: 'alexanderlebed999@gmail.com',
                password: 'user-password',
                pictureUrl: '',
                online: true,
                lastTime: 1519294933743,
                oauth: ''
            },
            {
                _id: '222',
                username: 'Alice',
                email: 'alice@gmail.com',
                password: 'alice-password',
                pictureUrl: '',
                online: true,
                lastTime: 1518346740388,
                oauth: ''
            },
            {
                _id: '333',
                username: 'Bob',
                email: 'bob@mail.com',
                password: 'bob-password',
                pictureUrl: '',
                online: true,
                lastTime: 1518346740388,
                oauth: ''
            }
        ],
        deleteUser: jest.fn(),
        translation: translation,
        history: { push: jest.fn() }
    };

    const targetUser = props.users[1];

    test('should match snapshot', () => {
        const wrapper = renderer.create(<People {...props} />);
        const tree = wrapper.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('should render people', () => {
        const wrapper = shallow(<People {...props} />);
        let userRows = wrapper.find(UserRow);
        expect(userRows.exists()).toEqual(true);
        expect(userRows.length).toEqual(3);
        expect(wrapper.instance().props.users).toEqual(props.users);
    });

    test('should alert confirmation on delete user', () => {
        const wrapper = mount(<People {...props} />);
        let confirmation = wrapper.find(ConfirmationModal);
        expect(confirmation.exists()).toEqual(false);
        expect(wrapper.instance().state.deleteUserId).toEqual('');

        const button = wrapper.find(`button#delete-user-${targetUser._id}`);
        expect(button.exists()).toEqual(true);
        button.simulate('click');
        confirmation = wrapper.find(ConfirmationModal);
        expect(confirmation.exists()).toEqual(true);
        expect(wrapper.instance().state.deleteUserId).toEqual(targetUser._id);
    });

    test('should route to conversation on \'write message\' click', () => {
        const spyPush = jest.spyOn(props.history, 'push');
        const wrapper = mount(<People {...props} />);
        expect(spyPush).toHaveBeenCalledTimes(0);

        const button = wrapper.find(`button#write-user-${targetUser._id}`);
        expect(button.exists()).toEqual(true);
        button.simulate('click');
        expect(spyPush).toHaveBeenCalledTimes(1);
        expect(spyPush).toHaveBeenLastCalledWith(`/conversation?userIds=${targetUser._id}`);
        spyPush.mockRestore();
    });

    test('should expand user image on image click', () => {
        const wrapper = mount(<People {...props} />);
        let userPictureModal = wrapper.find(`.profile-modal`);
        expect(userPictureModal.exists()).toEqual(false);

        const userPicture = wrapper.find(`img#user-picture-${targetUser._id}`);
        expect(userPicture.exists()).toEqual(true);
        userPicture.simulate('click');
        userPictureModal = wrapper.find(`.profile-modal`);
        expect(userPictureModal.exists()).toEqual(true);
    });
});