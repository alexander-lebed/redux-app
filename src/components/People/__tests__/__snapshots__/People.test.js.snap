// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<People/> should match snapshot 1`] = `
<div
  className="container-fluid"
>
  <div
    className="row"
  >
    <div
      className="col-lg-8 col-md-10 offset-lg-2 offset-md-1"
    >
      <div
        className="row"
      >
        <div
          className="col-sm-6 col-12"
        >
          <div
            className="search-container input-group input-group-sm"
          >
            <div
              className="input-group-prepend"
            >
              <i
                className="fas fa-search"
              />
            </div>
            <input
              aria-describedby="search"
              aria-label="Search people"
              className="search-input form-control"
              onChange={[Function]}
              placeholder="Search people"
              value=""
            />
          </div>
        </div>
      </div>
      <table
        className="table"
      >
        <tbody>
          <tr
            id="111"
          >
            <td
              style={
                Object {
                  "padding": 10,
                }
              }
            >
              <div
                className="row"
              >
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="profile-picture-wrapper cursor"
                    onClick={[Function]}
                  >
                    <img
                      className="profile-picture rounded-circle"
                      id="user-picture-111"
                      src="/images/default-profile.jpg"
                      style={
                        Object {
                          "boxShadow": "0px 0px 0pt 2pt #28a745",
                        }
                      }
                    />
                  </div>
                  <div>
                    Current User
                    <div
                      className="last-seen"
                    />
                  </div>
                </div>
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="pull-right pt-1 pt-sm-2"
                  >
                    <button
                      className="btn btn-outline-danger btn-sm"
                      disabled={false}
                      id="delete-user-111"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="far fa-trash-alt"
                        style={
                          Object {
                            "marginRight": 5,
                          }
                        }
                      />
                       
                    </button>
                     
                    <button
                      className="btn btn-outline-dark btn-sm"
                      disabled={false}
                      id="write-user-111"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="far fa-paper-plane"
                        style={
                          Object {
                            "marginRight": 5,
                          }
                        }
                      />
                       
                    </button>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr
            id="222"
          >
            <td
              style={
                Object {
                  "padding": 10,
                }
              }
            >
              <div
                className="row"
              >
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="profile-picture-wrapper cursor"
                    onClick={[Function]}
                  >
                    <img
                      className="profile-picture rounded-circle"
                      id="user-picture-222"
                      src="/images/default-profile.jpg"
                      style={
                        Object {
                          "boxShadow": "0px 0px 0pt 2pt #28a745",
                        }
                      }
                    />
                  </div>
                  <div>
                    Alice
                    <div
                      className="last-seen"
                    />
                  </div>
                </div>
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="pull-right pt-1 pt-sm-2"
                  >
                    <button
                      className="btn btn-outline-danger btn-sm"
                      disabled={false}
                      id="delete-user-222"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="far fa-trash-alt"
                        style={
                          Object {
                            "marginRight": 5,
                          }
                        }
                      />
                       
                    </button>
                     
                    <button
                      className="btn btn-outline-dark btn-sm"
                      disabled={false}
                      id="write-user-222"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="far fa-paper-plane"
                        style={
                          Object {
                            "marginRight": 5,
                          }
                        }
                      />
                       
                    </button>
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr
            id="333"
          >
            <td
              style={
                Object {
                  "padding": 10,
                }
              }
            >
              <div
                className="row"
              >
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="profile-picture-wrapper cursor"
                    onClick={[Function]}
                  >
                    <img
                      className="profile-picture rounded-circle"
                      id="user-picture-333"
                      src="/images/default-profile.jpg"
                      style={
                        Object {
                          "boxShadow": "0px 0px 0pt 2pt #28a745",
                        }
                      }
                    />
                  </div>
                  <div>
                    Bob
                    <div
                      className="last-seen"
                    />
                  </div>
                </div>
                <div
                  className="col-sm-6 col-12"
                >
                  <div
                    className="pull-right pt-1 pt-sm-2"
                  >
                    <button
                      className="btn btn-outline-danger btn-sm"
                      disabled={false}
                      id="delete-user-333"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="far fa-trash-alt"
                        style={
                          Object {
                            "marginRight": 5,
                          }
                        }
                      />
                       
                    </button>
                     
                    <button
                      className="btn btn-outline-dark btn-sm"
                      disabled={false}
                      id="write-user-333"
                      onClick={[Function]}
                      type="button"
                    >
                      <i
                        className="far fa-paper-plane"
                        style={
                          Object {
                            "marginRight": 5,
                          }
                        }
                      />
                       
                    </button>
                  </div>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      
    </div>
  </div>
</div>
`;
